### PR TITLE
[mlir][Transforms][NFC] Improve listener layering in dialect conversion

### PR DESCRIPTION
**Pull request by matthias-springer**
Context: Conversion patterns provide a `ConversionPatternRewriter` to modify the IR. `ConversionPatternRewriter` provides the public API. Most function calls are forwarded/handled by `ConversionPatternRewriterImpl`. The dialect conversion uses the listener infrastructure to get notified about op/block insertions.

In the current design, `ConversionPatternRewriter` inherits from both `PatternRewriter` and `Listener`. The conversion rewriter registers itself as a listener. This is problematic because listener functions such as `notifyOperationInserted` are now part of the public API and can be called from conversion patterns; that would bring the dialect conversion into an inconsistent state.

With this commit, `ConversionPatternRewriter` no longer inherits from `Listener`. Instead `ConversionPatternRewriterImpl` inherits from `Listener`. This removes the problematic public API and also simplifies the code a bit: block/op insertion notifications were previously forwarded to the `ConversionPatternRewriterImpl`. This is no longer needed.

### CODE REVIEW

**Comment by joker-eph** [comment]
> Review only the top commit.

What about using proper stacked PR by pushing your branch into the repo?

File: mlir/lib/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp [hunk 1]
```diff
@@ -582,7 +582,8 @@ class AwaitOpLoweringBase : public OpConversionPattern<AwaitType> {
     // Inside regular functions we use the blocking wait operation to wait for
     // the async object (token, value or group) to become available.
     if (!isInCoroutine) {
-      ImplicitLocOpBuilder builder(loc, op, &rewriter);
+      ImplicitLocOpBuilder builder(loc, rewriter);
+      builder.setInsertionPoint(op);
```
Comment (joker-eph) [inline mlir/lib/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp:586]: Why this change?

Comment (matthias-springer) [reply mlir/lib/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp:586]:
```c++
ImplicitLocOpBuilder builder(loc, op, &rewriter);
```
The above constructor sets the insertion point before `op` and attaches `&rewriter` as a listener.

After this commit, `ConversionPatternRewriter` no longer inherits from `Listener`, but we must still attach the same listener that is attached to `rewriter`. (`rewriter` used to have itself attached as a listener.) That's what `ImplicitLocOpBuilder builder(loc, rewriter);` does. There is no constructor that takes an existing `OpBuilder` (or subclass thereof) and op (to set the insertion point), so I have to manually set the insertion point with `setInsertionPoint`.

**Comment by matthias-springer** [comment]
> > Review only the top commit.
> 
> What about using proper stacked PR by pushing your branch into the repo?

Oh right, I'll give it a try on the next stack of commits.

File: mlir/lib/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp (hunk 1 above)
Comment (joker-eph) [reply mlir/lib/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp:586]: Isn't the rewriter already having the insertion point set? Why isn't `ImplicitLocOpBuilder builder(loc, rewriter);` taking the insertion point from there?

Comment (matthias-springer) [reply mlir/lib/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp]: You're right, this line is not needed. I didn't check what the insertion point of the rewriter was. It is also `op`.
